### ISSUE TITLE
Bug: Vulkan on Raspberry Pi 5 (V3D) — WG=512 / 16KB SMEM breaks inference

### ISSUE THREAD

**Issue by odilitime** [ggml-org/llama.cpp#26213]
### Summary
On Raspberry Pi 5 (V3DV Mesa), `llama.cpp` Vulkan either fails to create matmul pipelines (`Shared memory size too small for matrix multiplication`) or, after applying a low-SMEM tile path, produces **garbled** tokens unless several shaders that assume `local_size_x = 512` are adjusted for `maxComputeWorkGroupInvocations = 256`.

Related / prior: #9801 (closed stale), low-SMEM notes + patch from @davidscarth on that thread, interest from @geerlingguy.

### Hardware
- Raspberry Pi 5 16GB
- GPU: V3D 7.1.10.2 / Mesa 24.2.8 V3DV
- Limits (from ggml device log):
  - `maxComputeSharedMemorySize = 16384`
  - `maxComputeWorkGroupInvocations = 256`
  - `subgroupSize = 16`
  - no fp16 / int8 / integer dot / matrix cores

### Repro (stock master)
```bash
cmake -B build -DGGML_VULKAN=ON ...
./build/bin/llama-cli -m tinyllama.gguf -ngl 99 -p "Say hello" -n 16
```
Expected failure: shared-memory gate on mul_mat pipeline creation.

### Root causes (validated on b6176 + V3D)

1. **Matmul tiles need >16KB SMEM** on the default path. A gated low-SMEM warptile path (≤16KB) is required to even load layers (see patch discussed on #9801).

2. **Many elementwise / unary / GLU / GET_ROWS shaders hardcode `local_size_x = 512`**, while V3D only runs ≤256 invocations. Mesa does not cover the missing half of the workgroup:
   - Shaders that index with `y * 512 + x` (or `get_idx()` with that contract) but only launch 256 threads leave elements `[256..511]` unwritten per WG → compounding error → gibberish.
   - **ADD already uses the safe pattern**: `num_threads=256`, `num_iter=2`, `wg_denoms={512,1,1}`.
   - **RMS_NORM** uses `#define BLOCK_SIZE 512` + shared reduction → only half the lanes participate → NMSE≈1 in `test-backend-ops`.
   - **soft_max** specialization `{512}` exceeds device limit → must pass ≤256.
   - **GET_ROWS** uses plain `GlobalInvocationID.x` with `wg_denoms=512` + `local_size=512`; after capping local size to 256 without changing denoms, half of each embedding row is skipped. Unit tests use `n=256` and **miss** this.

3. Blindly changing all `wg_denoms` from `{512,1,1}` → `{256,1,1}` **breaks** ADD/`get_idx` (first mismatch at `i0=256`). Denoms and index stride must stay paired with the shader contract.

### Fixes that restored coherent English on Pi 5
- Low-SMEM matmul tile path (gated on `maxComputeSharedMemorySize <= 16384`).
- Align WG≤256 devices with the ADD pattern (256×2) for unary/GLU/copy-style shaders; keep `wg_denoms=512` where `get_idx` assumes 512.
- `RMS_NORM` / friends: `BLOCK_SIZE` ≤ device max (256 on V3D).
- `soft_max_*_wg512` specialization: pass `min(512, maxComputeWorkGroupInvocations)`.
- `GET_ROWS` / `ACC`: match `wg_denoms` to actual coverage (256 for 1-elem/thread; 512 for quant 2-elem/thread when local_size=256).

### Results (TinyLlama 1.1B Q4, temp 0)
| Mode | Coherent? | Decode |
|------|-----------|--------|
| CPU `-ngl 0` | yes | ~11–16 tok/s |
| Vulkan after fixes `-ngl 99` | yes (English; may diverge from CPU logits) | ~0.15 tok/s |

V3D is **not** a speed win vs CPU on Pi 5; this is a **correctness / enablement** fix so the backend does not silently corrupt state on WG≤256 / 16KB-SMEM devices.

### Ask
1. Reopen or replace #9801 with a maintained tracking issue.
2. Accept a PR that:
   - Gates a low-SMEM mul_mat path on device limits (no desktop regression).
   - Makes WG=512 shaders respect `maxComputeWorkGroupInvocations` (prefer specialization / ADD-style 256×2 rather than a global sed).
3. Add `test-backend-ops` cases with extents **>256** (e.g. 512, 2048) for GET_ROWS / unary / GLU so the half-coverage bug cannot regress.

Happy to iterate with `@ggml-org/ggml-vulkan`.

**Comment by odilitime** [comment]
PR ~~opened~~ closed for more discussion: https://github.com/ggml-org/llama.cpp/pull/26214
